src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs

```diff
@@ -93,53 +97,125 @@ public sealed class CosmosFhirDataStore : IFhirDataStore, IProvideCapability
-                        throw new ServiceUnavailableException();
+                if (weakETag != null && weakETag.VersionId != existingItemResource.Version)
+                {
+                    throw new PreconditionFailedException(string.Format(Core.Resources.ResourceVersionConflict, weakETag.VersionId));
                 }
 
-                _logger.LogError(exception, "Unhandled Document Client Exception");
+                if (existingItemResource.IsDeleted && cosmosWrapper.IsDeleted)
+                {
+                    return null;
+                }
 
-                throw;
+                existingItemResource.IsHistory = true;
+                existingItemResource.ActivePeriodEndDateTime = cosmosWrapper.LastModified;
+                existingItemResource.SearchIndices = null;
+
+                cosmosWrapper.Version = int.TryParse(existingItemResource.Version, out int existingVersion) ? (existingVersion + 1).ToString(CultureInfo.InvariantCulture) : Guid.NewGuid().ToString();
+
+                // indicate that the version in the raw resource's meta property does not reflect the actual version.
+                cosmosWrapper.RawResource.IsMetaSet = false;
+
+                if (cosmosWrapper.RawResource.Format == FhirResourceFormat.Json)
+                {
+                    // Update the raw resource based on the new version.
+                    // This is a lot faster than re-seserialing the POCO.
+                    // Unfortunately, we need to allocate a string, but at least it is reused for the HTTP response.
+
+                    // If the format is not XML, IsMetaSet will remain false and we will update the version when the resource is read.
+
+                    using MemoryStream memoryStream = _recyclableMemoryStreamManager.GetStream();
+                    await new RawResourceElement(cosmosWrapper).SerializeToStreamAsUtf8Json(memoryStream);
+                    memoryStream.Position = 0;
+                    using var reader = new StreamReader(memoryStream, Encoding.UTF8);
+                    cosmosWrapper.RawResource = new RawResource(reader.ReadToEnd(), FhirResourceFormat.Json, isMetaSet: true);
+                }
+
+                if (keepHistory)
+                {
+                    TransactionalBatchResponse transactionalBatchResponse = await _containerScope.Value.CreateTransactionalBatch(partitionKey)
+                        .ReplaceItem(cosmosWrapper.Id, cosmosWrapper, new TransactionalBatchItemRequestOptions { EnableContentResponseOnWrite = false, IfMatchEtag = existingItemResource.ETag })
+                        .CreateItem(existingItemResource, new TransactionalBatchItemRequestOptions { EnableContentResponseOnWrite = false })
+                        .ExecuteAsync(cancellationToken);
+
+                    if (!transactionalBatchResponse.IsSuccessStatusCode)
+                    {
+                        if (transactionalBatchResponse.StatusCode == HttpStatusCode.PreconditionFailed)
```

Comment:
Does the batch roll-up the status code? I think it also reports the status of each batch item

Reply:
Yes, this is a roll-up. Because there are only two operations, the status code will reflect the one that failed (the other's status will be dependency failed)